### ISSUE TITLE
corehost: Make shared framework and deps paths available to activated app

### ISSUE THREAD

**Issue by anurse** [dotnet/cli#1707]
In MVC and other contexts where we want to do dynamic compilation and loading, we re-read the deps file in managed code in order to properly identify all the necessary compilation/runtime dependencies. However, to fully resolve this, we need to do the same thing `corehost` does, namely merging this deps data with the one defined in the shared framework.

In the current design, `corehost` simply uses it's location to determine where the shared framework lives (i.e. in muxer mode, it reads the framework name and version and looks in `OWNDIR/../shared/[name]/[version]`).

Similarly, when corehost is invoked with `--deps`, we want our managed APIs to be able to identify the path to that deps file. This is needed for xunit to be able to load properly. /cc @piotrpMSFT 

We need to look for a way for `corehost` to provide this data to the managed application. Perhaps through a P/Invoke (managed app P/Invokes back into corehost?) or other shared state? Environment variables would also be a simple way to do this for now (we did this with DNX and it was somewhat problematic, but did work)

/cc @gkhanna79 @schellap @pakrym @davidfowl

**Comment by pakrym** [comment]
This wold become blocker for MVC when we move to "fat" apps

**Comment by anurse** [comment]
Ping @gkhanna79 @schellap @davidfowl @piotrpMSFT @Petermarcu 

This is going to block MVC. Since the only one who knows the location of the shared framework and the deps file are is corehost, we need to try and come up some something to propagate that data to managed code so we can properly load additional assemblies or create compilations based on that data.

There are some hacks we could use, but none of them cover anything, and my hope is that since `corehost` has this data readily available, we could just stuff it in environment variables or something.

**Comment by schellap** [comment]
We can have an util function exported out of a global storage for this. Or as a property on appdomain appcontext similar to basedirectory.

**Comment by anurse** [comment]
Not sure that a property off the AppContext would be accessible to managed code, since I think those are not exported in CoreCLR.

A util function would be OK. We'd have to P/Invoke back into corehost to get it, but it would be fairly safe.

**Comment by schellap** [comment]
What is the exact info you need? Paths to the fx dir and app dir isn't sufficient? You should be able to load up the deps from there.

**Comment by schellap** [comment]
> would be accessible to managed code, since I think those are not exported in CoreCLR

@weshaggard, is it non trivial to expose this if set as a property to the CLR like app context base?

**Comment by anurse** [comment]
So something like the following would probably be ideal:

```c++
wchar_t* corehost_get_deps_file_list()
```

Note: Since we want to read this from managed code, I assume we'll need it to return `wchar_t`.

Which we could P/Invoke with:

```csharp
[DllImport("hostpolicy.dll?")]
public unsafe extern char* corehost_get_deps_file_list()
```

The value could just be an **ordered** (we need to know which was applied first, second, etc. so we can figure out which is the shared framework one, etc.) `PATH`-separator (`;` on Windows, `:` on Unix) delimited list of paths to deps files.

Thoughts @davidfowl @pakrym ? It's not pretty to have to P/Invoke, but the alternative is basically to have an environment variable (which has it's own problems due to inheritance of the env vars).

**Comment by gkhanna79** [comment]
PInvoking to the host to get this data is a bad design and dependency. The host is an implementation detail that managed code should be unaware of this. It should be free to change underneath or even disappear (e.g. In CoreRT, the host maybe different artifact, or incases, none, at all).

> AppContext doesn't have any of this information right now and no generic data bag to put it into either.

@weshaggard , currently AppDomain initialization gets the Key/Value pairs from the host. For CLI, these are initialized in https://github.com/dotnet/cli/blob/328e806069252f3fff3854c9c451d98ae71100da/src/corehost/cli/hostpolicy.cpp and passed for AppDomain initialization where other things like AppContext Compat mode, etc get passed. Is there a reason these values cannot be passed in the same manner to the AppDomain initialization and AppContext can pick it from there for its initialization and make it available for others to consume?

**Comment by davidfowl** [comment]
I,d prefer that actually. As long as we can round trip the data to managed code properly

**Comment by schellap** [comment]
@anurse, just a thought, would a combination of `dladdr(coreclr.dll)`, similar for windows in conjunction with `BaseDirectory`, give you the two directories you need and then getting the deps file is based on `GetExecutingAssembly`.deps.json and `netcoreapp.deps.json` if coreclr dir is different from basedir?

Of course app will break if the app activation model changes, but that would be rare.

**Comment by davidfowl** [comment]
Here's the library:

https://github.com/dotnet/cli/tree/rel/1.0.0/src/Microsoft.Extensions.DependencyModel

This is the code that needs to do the right thing and get both deps file so that applications can have a full view of dependencies.

**Comment by pakrym** [comment]
@schellap we can not get location passed via `--deps` corehost argument this way.

**Comment by gkhanna79** [comment]
@weshaggard What are your thoughts on AppContext delivering these values based upon my suggestion earlier?

**Comment by schellap** [comment]
@pakrym `--depsfile` arg to the host will set the appcontext base to app location. GetExecutingAssembly at app entry should give you app name or you can just hard code this in the app to. This should give you the deps file. I don't think you would ever rename the deps file and pass it to --depsfile.

Coreclr should givr you fx dir for Netcoreapp.deps.json.

**Comment by pakrym** [comment]
>  I don't think you would ever rename (EDIT: move) the deps file and pass it to --depsfile
We are passing test library deps file along with `xunit.runner.dll` as entry point dll.

**Comment by schellap** [comment]
`We are passing test library deps file along with xunit.runner.dll as entry point dll.`

Don't follow where this breaks, could you please let me know who will fail to know about the deps file? The app being tested? Or xunit?

You still have the deps file's directory at appcontext basedirectory. The app can get to it using its own name.

**Comment by anurse** [comment]
I am totally in favor of a solution that involves passing the data through `AppContext` rather than a P/Invoke. The P/Invoke is absolutely not great layering, I'm just trying to get us a solution.

At the core, we need to know about all the deps files that were used. In the standard scenario, we can use `typeof(object).Assembly.Location` to get the one next to `mscorlib` (shared framework) and `AppContext.BaseDirectory` to get the one next to the app. However, in tests and `dotnet-[toolname]` activations (/cc @piotrpMSFT), the deps file is **not** co-located with the entry-point assembly, it is passed in via the `--depsfile` switch to corehost.

**Even if** corehost were to ensure the `AppContext.BaseDirectory` is pointing at the location containing the deps file, we don't know the **name** of the deps file since it is derived from the test assembly name, which is **not** the same as the entry assembly name (which is the xunit runner).

**Comment by gkhanna79** [comment]
> we don't know the name of the deps file since it is derived from the test assembly name, which is not the same as the entry assembly name (which is the xunit runner).

The host can pass the paths to the respective .deps.json and AppContext exposes them.

**Comment by anurse** [comment]
Yes, that would be ideal. I was just clarifying that we do not currently have enough data for that today and would need more information to be flowed up to managed code.

**Comment by weshaggard** [comment]
OK I'm going to get an API added to AppContext to be used for this. see https://github.com/dotnet/corefx/issues/6871. For the immediate unblocking please use Environment variables as workaround.

**Comment by weshaggard** [comment]
To close the door on this issue once AppContext.GetData has be added so any package update over the last week or so should have that API.

**Comment by livarcocc** [comment]
@anurse Now that this API is available, can you check that that's enough for you and close this issue?

**Comment by anurse** [comment]
Should be enough, but we need to convert DependencyContext to use it. /cc @pakrym

**Comment by anurse** [comment]
AFAIK, we still need to convert DependencyContext to use the new API. It's using reflection right now

**Comment by gkhanna79** [comment]
Reopened based upon @anurse comment above.

**Comment by anurse** [comment]
This should go in RTM though, since it's not blocking a user scenario. /cc @piotrpMSFT @blackdwarf

**Comment by piotrpMSFT** [comment]
Sounds good @anurse. Looking forward to the PR.

**Comment by pakrym** [comment]
Fixed with #2757
